User LO setting independent of Schedule and assembler update

itoko opened the pull request:
This PR address the issue that `Schedule` is not reusable when changing lo freqs for each experiments.
- introduces `ScheduleConfig` class to manage user's lo configuration
- enables to avoid from unnecessary creation of `Schedule`
- make `OutputChannel` immutable

For example, 
```
[Current]
for drive_lo_freq in drive_lo_freqs:
    freq_d0 = freq_d0_default + drive_lo_freq

    # compose schedule
    schedule = pulse.Schedule()
    schedule.insert(0, rabi_pulse(device.q[0].drive))
    schedule.insert(drive_samples, meas_pulse(device.q[0].measure))
    schedule.insert(drive_samples, pulse.Acquire(duration=meas_samples)(device.q[0].acquire))

    # set LO frequency
    device.q[0].drive.set_lo_frequency(freq_d0)
    schedules.append(schedule)

execute_schedules(schedules, backend, **opts)
```
```
[Proposed (1 schedule : n lo_configs)]
# compose schedule
schedule = pulse.Schedule()
schedule.insert(0, rabi_pulse(device.q[0].drive))
schedule.insert(drive_samples, meas_pulse(device.q[0].measure))
schedule.insert(drive_samples, pulse.Acquire(duration=meas_samples)(device.q[0].acquire))

# set LO frequency
lo_configs=[]
for drive_lo_freq in drive_lo_freqs:
    freq_d0 = freq_d0_default + drive_lo_freq
    lo_configs.append(ScheduleConfig(user_lo_dic={device.q[0].drive: freq_d0}))

execute_schedules(schedule, backend, lo_configs, **opts)
```
```
[Proposed (n schedules : 1 lo_config)]
# compose to schedules
schedules = []
for ii in pulses:
    schedule = pulse.Schedule()
    schedule.insert(0, pulses[ii](device.q[0].drive))
    schedule.insert(drive_samples, meas_pulse(device.q[0].measure))
    schedule.insert(drive_samples, pulse.Acquire(duration=meas_samples)(device.q[0].acquire))
    schedules.append(schedule)

# set LO frequency
lo_config = ScheduleConfig(user_lo_dic={device.q[0].drive: freq_d0})

execute_schedules(schedules, backend, lo_config, **opts)
```

For users who use default los, no change in interface:
```
execute_schedules(schedule, backend, **opts)
```

itoko committed:
change user-lo setting interface

itoko committed:
adapt execute/assembler for new user-lo setting interface

itoko committed:
fix tiny bugs

nkanazawa1989 commented:
@itoko thank you for the quick work! I have several comments.

The important point is that a user can specify `LoConfig` for arbitrary schedule. In this PR, user is forced to use the same `Schedule` for LO configuration.
I think this limitation should be relaxed, because sometime we want to change both schedule and frequency in the single job. Currently quantum computers are not stable and default frequency is constantly updated by calibration. Thus, combination of results from separated jobs may give us unexpected frequency detuning. It deteriorates accuracy of experiment.

nkanazawa1989 commented on qiskit/compiler/assembler.py:
```diff
@@ -142,40 +142,31 @@ def assemble_schedules(schedules, dict_config, dict_header):
+    for ii, configed in enumerate(schedules):
+        schedule = configed.schedule
+
+        lo_freqs = {
```
Please avoid adding them if default and user_configs are the same. Unnecessary header configs just increase data size.

itoko replied:
OK. In the case of default_los == user_los, we need to set `lo_freqs = {}`, right?

itoko replied:
done!

nkanazawa1989 reviewed (changes requested):
@itoko thank you for the quick work! I have several comments.

The important point is that a user can specify LoConfig for arbitrary schedule. In this PR, user is forced to use the same Schedule for LO configuration.
I think this limitation should be relaxed, because sometime we want to change both schedule and frequency in the single job. Currently quantum computers are not stable and default frequency is constantly updated by calibration. Thus, combination of results from separated jobs may give us unexpected frequency detuning. It deteriorates accuracy of experiment.

nkanazawa1989 commented:
Another point is the returned value of `LoConfig`. If we align `TranspileConfig` , we need to prepare marshmallow schema and return `QobjLoConfig`.

chriseclectic committed:
fix numpy casting error with *= (#2062)

taalexander commented:
Yes I agree with Naoki. Its possible we require something like a `ScheduleConfig` which are allowed to be supplied for each experiment. The lo could be a parameter of this.

nonhermitian committed:
Allow for list of ints as layout (#2058)

* int list as layout

* make integer layout a method of Layout

* integer_layout -> intlist_layout

* similar interface to generate_trivial_layout

* move the test to test_compile

* test for the errors

* changelog

* If the intlist is longer that your quantum register, map them to None.

* valid names

* the name should make it obvious that constructs an object

* lint

itoko committed:
Merge branch 'commands-rework-assembler' into lo-rework

itoko commented:
I'll do two improvements:
- move LoConfig behind of `ScheduleConfig`
- handle 1:1, 1:n, n:1, n:n case for `#Schedule`:`#LoConfig`
I think we don't need to rush to align with marshmallow at this time since it's not the issue of user interface.

diego-plan9 committed:
Remove underscore from quantum_info, providers (#2060)

* Remove underscore from basicaertools

* Remove underscore from quantum_info

nonhermitian committed:
cast to uint32 (#2063)

itoko committed:
move lo config under ScheduleConfig

itoko committed:
improve to work when #schedules:#configs = n:n, 1:n or n:1

itoko committed:
improve not to create unnecessary lo_freqs

nonhermitian committed:
Allow for bit string distance measures (Hamming dist) in plot_histogram (#2064)

* allow for sorting histograms by hamming dist

* lint

* Update qiskit/tools/visualization/_counts_visualization.py

Co-Authored-By: nonhermitian <nonhermitian@gmail.com>

* add to changelog

* fix spelling

ajavadia committed:
Remove qargs/cargs/circuit from instructions, in favor of circuit keeping context (#1816)

* remove qarg and carg from Instruction, modify circuit._attach

* update circuit_to_dag, remove qarg/carg from all gates

* fixing decomposition rule definitions

* update to decomposition rules

* bring measure qasm under QuantumCircuit

* add args context to InstructionSet

* fix extension tests, update matrix unitary gate to no qargs

* update snapshot instruction

* fix barrier insertion pass

* update transpiler passes for new format, rest of decomp rules

* add instruction shapes (num_qubits, num_clbits)

add instruction shapes (num_qubits, num_clbits)

* fix tests except initialize

* initialize is a generic instruction with decomposition rules

* move decompositions to generic instructions, add circuit_to_instruction

* add circuit.append()

* generalize dag_to_circuit

* use copy in circuit_to_dag and dag_to_circuit, temporary slowdown

* allow unroller pass to unroll generic instructions, not just gates

* initialize as Instruction that recursively decomposes

* revert to only gates being unrolled, and method for circuit_to_gate for now

* make qargs and cargs optional in circuit.append and dag.apply_operation_back

* add circuit.to_gate() method to wrap the convertor

* cleanup initialize

* use a list of Gate for a Gate.definition, decompose based on that

* add flag to indicate instruction reversibility

* decompose pass does not error on fundamental gates

* rename init to initialize in qobj

* circuit.decompose() method

* remove instruction._modifiers method

* remove instruction.circuit property

* CHANGELOG

* allow general subroutines via instruction definitions and update unroller

* instruction.reverse() -> instruction.mirror()

* inplace option for Instruction mirror and inverse

* base gates should not modify with inverse. add tests

* flatten register in circuit_to_instruction conversion

jwoehr committed:
Add quiet and output to job monitor to suppress or redirect status messages (#2037)

* job monitor takes quiet= (bool:default False) and to_file = (file:default None == sys.stdout)

* Update qiskit/tools/monitor/job_monitor.py

Co-Authored-By: jwoehr <4604036+jwoehr@users.noreply.github.com>

* Update qiskit/tools/monitor/job_monitor.py

Co-Authored-By: jwoehr <4604036+jwoehr@users.noreply.github.com>

* Update qiskit/tools/monitor/job_monitor.py

Co-Authored-By: jwoehr <4604036+jwoehr@users.noreply.github.com>

* Update qiskit/tools/monitor/job_monitor.py

Co-Authored-By: jwoehr <4604036+jwoehr@users.noreply.github.com>

* Update qiskit/tools/monitor/job_monitor.py

Co-Authored-By: jwoehr <4604036+jwoehr@users.noreply.github.com>

* Update qiskit/tools/monitor/job_monitor.py

Co-Authored-By: jwoehr <4604036+jwoehr@users.noreply.github.com>

* Update qiskit/tools/monitor/job_monitor.py

Co-Authored-By: jwoehr <4604036+jwoehr@users.noreply.github.com>

* Update qiskit/tools/monitor/job_monitor.py

Co-Authored-By: jwoehr <4604036+jwoehr@users.noreply.github.com>

* Update qiskit/tools/monitor/job_monitor.py

Co-Authored-By: jwoehr <4604036+jwoehr@users.noreply.github.com>

* Update qiskit/tools/monitor/job_monitor.py

Co-Authored-By: jwoehr <4604036+jwoehr@users.noreply.github.com>

* broke too-long lines that failed pylint

chriseclectic committed:
fix sdg definition (#2069)

maddy-tod committed:
Fixed mpl drawer output (#2061)

* Updated the mpl drawer so that the output is no longer messy

* Renamed gate_index to gate_anchor

* Fixed off by 1 error - gates now begin in column 0 not 1

* Fixed linting

nonhermitian committed:
Return minimal bandwidth best layout (#2021)

* Compute a minimal bandwidth best layout

* flip cx gates in offending test

* flipped wrong test

* update changelog

* fixed tests

nkanazawa1989 commented on qiskit/execute.py:
```diff
@@ -158,6 +161,23 @@ def execute_schedules(schedules, backend, **kwargs):
+    if schedule_configs:
+        if isinstance(schedule_configs, ScheduleConfig):
+            schedule_configs = [schedule_configs]
+        if len(schedules) == len(schedule_configs):
```
Can this be like
```
for schedule_config in schedule_configs:
    for schedule in schedules:
        experiments = ConfiguredSchedule(schedule, schedule_config)
```
In this way users can easily generate a recipe of spectrum mapping.

nkanazawa1989 replied:
for n:n case, we can directly provide `ConfiguredSchedule`.

nkanazawa1989 commented on qiskit/pulse/schedule/experiment_config.py:
```diff
@@ -0,0 +1,42 @@
+
+class ScheduleConfig:
+    """Configurations for user LO frequencies."""
+
```
Can you add `Example::` how users can construct `ScheduleConfig` ? Specification of `user_lo_dic` is not intuitive.

nkanazawa1989 commented on qiskit/execute.py:
```diff
@@ -19,8 +19,10 @@
 from qiskit.compiler import RunConfig, TranspileConfig
+from qiskit.compiler import assemble_circuits, assemble_schedules, transpile
+from qiskit.exceptions import QiskitError
+from qiskit.pulse import Schedule, ConfiguredSchedule, ScheduleConfig
```
Can you move `ScheduleConfig` to `qiskit.compiler.schedule_config` ? Then we can import it as
```
from qiskit.compiler import RunConfig, TranspileConfig, ScheduleConfig
```
This is more aligned with how `QuantumCircuit` is assembled.

nkanazawa1989 commented:
@itoko Thank you for the updates! That is great. I want to add small changes and this can be merged. Basically I agree with 
> I think we don't need to rush to align with marshmallow at this moment since it's not the issue of user interface.

itoko committed:
add a reference schedule for pulse

itoko committed:
add a test for assembler

itoko committed:
make sample pulses hashable

nkanazawa1989 commented:
Can we also remove `lo_frequency` from channels?

itoko committed:
fix a bug in Acquire constructor with default values

itoko committed:
fix bugs in assembler and test_assembler

itoko committed:
simplify reference schedule

itoko committed:
refactor meas_opts

nkanazawa1989 committed:
Merge pull request #22 from nkanazawa1989/assember-test

Add a test for assemble_schedule

Commit:
fix merge error

Commit:
fix wrong field name

Commit:
add test for command name

itoko committed:
Merge pull request #25 from nkanazawa1989/fix_merge

fix merge error

itoko commented:
@nkanazawa1989 I think we still need to get `lo_frequency`. Maybe we should rename it to `default_lo`. I guess users may want to do like `user_lo = device.q[0].drive.lo_frequency + 0.01`.

ajavadia committed:
test unrolling for all gates (#2072)

mtreinish committed:
Just use a list not split a string (#2071)

In the matplotlib circuit drawer backend there was a statically defined
list of gates to treat as wide. For some reason this list was defined by
build a space separated string of the gates and then running the split()
method on that string. This is unecessary and we can just define a list
of these gates. This commit does just that and defines those gates as a
list and removes the split() uasge.

1ucian0 committed:
remove_op_node is part of the DAG API (#2024)

* remove_op_node is part of the DAG API

* lint

* docstirng

* raises in case of non node.type != op

* test_remove_op_node_longer

* Update qiskit/dagcircuit/dagcircuit.py

Co-Authored-By: 1ucian0 <luciano@debian.org>

* new apply_operation_back
api

ewinston committed:
Add arbitrary unitary gate (#1961)

* start unitary class and add some tests

tests passing

insert Unitary in QuantumCircuit

minor commit

add power

linting, bug fix

added tests

remove check_qubit during reapply

Extend InstructionParameter for containers

Extend `InstructionParameter`, adding support for containers (`list`
and `numpy.ndarray`. Add tests for the field.

Resync elapsed time with ISO 80000-3 annex

fixes bug (uses PR #1970)

fix silent bad type

* linting

* typo

* remove matrix method. Add _representation method

* bug fixes

* fix init with Unitary. Overload __eq__.

* minor fix

* adapt to pr #816 removing qargs from instructions.

* minor linting

* remove reapply and decomposition methods

these methods are not longer needed after recent upstream changes.

* remove unused import

chriseclectic committed:
fix occasionally failing tests from non-unique Kraus conversion (#2075)

mtreinish committed:
Migrate visualization to top level namespace (#1878)

* Migrate visualization to top level namespace

This commit moves the visualization code out of tools into it's own top
level namespace, 'qiskit.visualization'. The visualization was a bit out
of place in the tools directory and is really it's own top level thing
at this point. However, because we declared the public visualization
interfaces stable as part of the 0.7 release we must maintain backwards
compatibility. In order to facilitate that all the public APIs are still
accessible via 'qiskit.tools.visualization'. However, private
api/modules are not accessible off of 'qiskit.tools.visualization'
anymore.

* Add changelog

* Fix lint

Sigh pylint, raising an error for both using a wildcard import and not
using all the imports from the wild card.

* Rebase updates

* Move tests to match new dir structure

* Update test_visualization import per review comment

* Remove underscores from visualization modules

* Fix lint

1ucian0 committed:
ResourceEstimation AnalysisPass (#1889)

* count operations pass

* empty dag test

* rename

* test

* test rename

* test name

* Update qiskit/transpiler/passes/resource_estimation.py

Co-Authored-By: 1ucian0 <luciano@debian.org>

* count_ops

* count_ops

* lint

* changelog

* depth pass

* size pass

* size pass

* count_ops pass

* size pass

* NumTensorFactor pass

* resource calculation metapass

* lint

* unused import

* changelog

* lint

* Update qiskit/transpiler/passes/__init__.py

Co-Authored-By: 1ucian0 <luciano@debian.org>

* Update qiskit/transpiler/passes/width.py

Co-Authored-By: 1ucian0 <luciano@debian.org>

* Update qiskit/transpiler/passes/num_tensor_factor.py

Co-Authored-By: 1ucian0 <luciano@debian.org>

* Update qiskit/transpiler/passes/depth.py

Co-Authored-By: 1ucian0 <luciano@debian.org>

* Update qiskit/transpiler/passes/depth.py

Co-Authored-By: 1ucian0 <luciano@debian.org>

* Update qiskit/transpiler/passes/size.py

Co-Authored-By: 1ucian0 <luciano@debian.org>

* Update qiskit/transpiler/passes/width.py

Co-Authored-By: 1ucian0 <luciano@debian.org>

* Update qiskit/transpiler/passes/size.py

Co-Authored-By: 1ucian0 <luciano@debian.org>

* Update qiskit/transpiler/passes/num_tensor_factor.py

Co-Authored-By: 1ucian0 <luciano@debian.org>

* renaming

mtreinish committed:
Don't pass an empty RunConfig to assemble_circuits (#2011)

* Don't pass an empty RunConfig to assemble_circuits

This commit modifies the deprecated circuits_to_qobj() converter
function to handle the case where the run_config kwarg is not set
properly. Previously the function was initializing an empty RunConfig
option unecessarily and passing that into assemble_circuits(). However,
this was not necessary because assemble_circuits() can handle the case
where run_config is not set.  So this just removes that and passes None
directly to assemble_circuits() which fixes potential validation errors.

Fixes #1988

* Fix lint

* More lint fixes

Commit:
Merge branch 'master' of github.com:Qiskit/qiskit-terra into lo-rework
